File: frontend/src/types/broker-rules.ts
/**
 * Broker rule matrix types — mirrors the backend `domain.BrokerOrderRule`
 * (Phase 5). Consumed by the capability-native order ticket
 * (PlaceOrderDialogV2).
 */

import type { AssetClass, PlatformOrderType, PlatformQuantityUnit, PlatformSession, PlatformTimeInForce } from './capabilities'

export type OrderSide = 'BUY' | 'SELL'

export interface BrokerOrderRule {
  venue_code: string | null
  asset_class: AssetClass | null
  session: PlatformSession | null
  side: OrderSide | null
  quantity_unit: PlatformQuantityUnit | null
  allowed_order_types: PlatformOrderType[]
  allowed_time_in_force: PlatformTimeInForce[]
  allows_fractional: boolean
  allows_notional: boolean
  allows_short: boolean
  requires_limit_price: boolean
}

/** Count of non-null qualifier fields — higher is more specific. */
export function ruleSpecificity(rule: BrokerOrderRule): number {
  let score = 0
  if (rule.venue_code !== null) score++
  if (rule.asset_class !== null) score++
  if (rule.session !== null) score++
  if (rule.side !== null) score++
  if (rule.quantity_unit !== null) score++
  return score
}

/** Pick the most specific rule matching ``(venue, assetClass)``. */
export function pickMatchingRule(
  rules: BrokerOrderRule[],
  venueCode: string | null,
  assetClass: AssetClass | null,
): BrokerOrderRule | null {
  const matches = rules.filter((r) => {
    if (r.venue_code !== null && r.venue_code !== venueCode) return false
    if (r.asset_class !== null && r.asset_class !== assetClass) return false
    return true
  })
  if (matches.length === 0) return null
  matches.sort((a, b) => ruleSpecificity(b) - ruleSpecificity(a))
  return matches[0]
}

File: domain/broker_rules.py
"""BrokerOrderRule — declarative per-broker order constraints.

A Pydantic mirror of :class:`database.broker_rules_repo.BrokerOrderRulesRow`.
Rows are loaded from the DB, wrapped into this model for validation,
and consumed by :mod:`services.rule_enforcement`.
"""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, ConfigDict

from domain.enums import AssetClass, OrderSide, OrderType, QuantityUnit, Session, TimeInForce


class BrokerOrderRule(BaseModel):
    """Declarative rule describing what a broker allows.

    Rules with more non-null qualifier fields are considered more
    specific. The rule-enforcement service picks the most specific rule
    that matches the order.
    """

    model_config = ConfigDict(frozen=True, extra="forbid")

    broker_code: str
    venue_code: str | None = None
    asset_class: AssetClass | None = None
    session: Session | None = None
    side: OrderSide | None = None
    quantity_unit: QuantityUnit | None = None
    allowed_order_types: list[OrderType]
    allowed_time_in_force: list[TimeInForce]
    requires_limit_price: bool = False
    allows_fractional: bool = False
    allows_notional: bool = False
    allows_short: bool = False
    metadata: dict[str, Any] = {}

    @property
    def specificity(self) -> int:
        """Number of non-null qualifier fields. Ties are broken by the
        caller's preferred order (more specific first)."""
        return sum(
            1
            for v in (
                self.venue_code,
                self.asset_class,
                self.session,
                self.side,
                self.quantity_unit,
            )
            if v is not None
        )


__all__ = ["BrokerOrderRule"]
